src/crimson/os/seastore/btree/fixed_kv_node.h

```diff
@@ -27,29 +28,716 @@ namespace crimson::os::seastore {
  * Base class enabling recursive lookup between internal and leaf nodes.
  */
 template <typename node_key_t>
-struct FixedKVNode : CachedExtent {
+struct FixedKVNode : ChildableCachedExtent {
+
+  enum op_t : uint8_t {
+    NONE = 0,
+    INSERT,
+    REMOVE,
+    UPDATE,
+    NUM_OPS
+  };
+
+  // this structure is for efficient merge of child poiner arrays
+  struct pending_child_tracker_t
+    : public boost::intrusive::set_base_hook<
+	      boost::intrusive::link_mode<
+		boost::intrusive::auto_unlink>>
+  {
+    ChildableCachedExtent* child = nullptr;
+    uint16_t pos = 0;
+    node_key_t key;
+    op_t op = NONE;
+
+    pending_child_tracker_t(
+      ChildableCachedExtent* child,
+      uint16_t pos,
+      node_key_t key,
+      op_t op)
+      : child(child),
+	pos(pos),
+	key(key),
+	op(op)
+    {}
+
+    pending_child_tracker_t(const pending_child_tracker_t &other)
+      : child(other.child),
+	pos(other.pos),
+	key(other.key),
+	op(other.op)
+    {}
+
+    friend bool operator<(
+      const pending_child_tracker_t &left,
+      const pending_child_tracker_t &right) {
+      return left.key < right.key;
+    }
+    friend bool operator>(
+      const pending_child_tracker_t &left,
+      const pending_child_tracker_t &right) {
+      return left.key > right.key;
+    }
+    friend bool operator==(
+      const pending_child_tracker_t &left,
+      const pending_child_tracker_t &right) {
+      return left.key == right.key;
+    }
+
+    struct cmp_t {
+      bool operator()(
+	const pending_child_tracker_t &left,
+	const node_key_t &right) const {
+	return left.key < right;
+      }
+      bool operator()(
+	const node_key_t &left,
+	const pending_child_tracker_t &right) const {
+	return left < right.key;
+      }
+    };
+  };
+
   using FixedKVNodeRef = TCachedExtentRef<FixedKVNode>;
+  using pending_children_set_t =
+    boost::intrusive::set<
+      pending_child_tracker_t,
+      boost::intrusive::constant_time_size<false>>;
+
+  fixed_kv_node_meta_t<node_key_t> range;
+
+  /*
+   *
+   * mutate_state_t
```

Comment:
mutate_state_t and pending_child_tracker_t aren't in the design we discussed before.  Why can't we update each child as we invalidate the old one as discussed?  This seems to break the invariant that extents only have bare pointers to extents that hold strong references back.

In a larger sense, the commit "crimson/os/seastore/btree: introduce parent<->child pointers for fixed-kv-btree nodes" is too complicated to review without adding an explanation of the invariants you are maintaining.  Please add something like the rules I give in https://github.com/ceph/ceph/pull/47749#issuecomment-1260289249.

Reply:
> mutate_state_t and pending_child_tracker_t aren't in the design we discussed before. Why can't we update the parent from each child as it is invalidated as discussed?

`mutate_state_t` is used to reference `initial_pending` children in `mutation/initial_pending` parents. `mutate_state_t` fields in `non-mutation/initial_pending` nodes are empty. `pending_child_tracker_t` and `mutate_state_t::pending_insert/update/remove/replace` are introduced to avoid the unnecessary memory copies in the following scenario:

1. A stable node `A` has 100 children;
2. A new `C1` child is inserted to the `mutation_pending` version of `A` at pos 0, which would lead to a position shifting of pointers from pos 1 to 100;
3. Another `C2` is inserted at pos `0`, which again shifts pointers from pos 1 to 101;

In this case the first pointers positions shifting is unnecessary, so I introduced the `pending_child_tracker_t`. In general, to avoid the unnecessary memory copies in the above case, `mutation_state_t::pending_insert/remove/update/replace` methods only shift the would-be pos of pending children.

> This seems to break the invariant that extents only have bare pointers to extents that hold strong references back.

The invariants this PR holds are not changed, still the ones in https://github.com/ceph/ceph/pull/47749#issuecomment-1260289249, I think the code is somewhat misleading, will try to cleanup the code and add more comments about `mutate_state_t` and `pending_child_tracker_t`:-)

Reply:
To clarify, there needs to be a comment within this class explaining the invariants even if the comment is identical to https://github.com/ceph/ceph/pull/47749#issuecomment-1260289249.  A new contributor in a month or two needs to be able to find out how these structures work without needing to dig through the github comments on some old PR.

Reply:
I'm skeptical that dynamically allocating pending_child_tracker_t structures is actually a performance win compared with an extra memmove in the event that the pending parent happens to have more than once pending child.  It also strikes me as quite complex for a relatively narrow case.

Reply:
> To clarify, there needs to be a comment within this class explaining the invariants even if the comment is identical to https://github.com/ceph/ceph/pull/47749#issuecomment-1260289249.  A new contributor in a month or two needs to be able to find out how these structures work without needing to dig through the github comments on some old PR.

Yeah, will add it.

Reply:
> I'm skeptical that dynamically allocating pending_child_tracker_t structures is actually a performance win compared with an extra memmove in the event that the pending parent happens to have more than once pending child.  It also strikes me as quite complex for a relatively narrow case.

I understand that, will try to remove it.

Reply:
@athanatos Hi, sam, when trying to remove `mutate_state_t`, I recalled that there were another reason for adding it.

The invariants in https://github.com/ceph/ceph/pull/47749#issuecomment-1260289249 require that `mutation/initial_pending` fixed-kv nodes only hold pointers to `initial_pending` children, other pointers are copied from `copy sources` when committing `pending` fixed-kv nodes.

This requires that:

1. for updates/inserts, we always keep the positions of `initial_pending` children somewhere, so that we know the new position of those copied pointers when copying pointers from `copy sources`;
2. for removals, we must keep those removals' keys and positions until committing, so that we know which pointer shouldn't be copied.

`mutate_state_t` actually serves as that in-mem store of `pending positions`. Without `mutate_state_t`, we'll probably have to rely on comparing each and every Key-Value pairs in the `pending` fixed-kv nodes with those in the `copy sources` to get correct new `stable` fixed-kv nodes. The time complexity of this comparison is probably `O(m + n)` with `m` being the number of kv pairs in the `pending` fixed-kv node and `n` being that in `copy sources`. This comparison happens each time we commit a `pending` fixed-kv node.

@athanatos Do you still think we should drop this `mutate_state_t`? Or, maybe I can cleanup the code, so it would be more readable.

Reply:
1. Keep them in the corresponding position in children.  On commit, do a linear (yes, O(m+n)) merge of the two sets.  We can improve this later.
2. Nope, the above linear scan can simply notice that the removed child doesn't exist on the new node.

Reply:
Understood:-)